File: JULY-2025/17-06-1015/p3202.php
class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $k
     * @return Integer
     */
    function maximumLength($nums, $k) {
        // Create a 2D array to store the frequency of valid subsequence of length i,
        // ending at index j with the remainder of (nums[$i] + nums[$j]) % $k as the key.
        $dp = array_fill(0, count($nums), array_fill(0, $k, 0));

        // Initialize the maximum length of valid subsequence to 0.
        $maximum = 0;

        // Iterate over the indices in reverse order.
        for ($i = count($nums) - 1; $i >= 0; $i--) {
            for ($j = $i + 1; $j < count($nums); $j++) {
                // Calculate the remainder of (nums[$i] + nums[$j]) % $k.
                $remainder = ($nums[$i] + $nums[$j]) % $k;
                // Increment the frequency of the valid subsequence by 1.
                $frequency = $dp[$j][$remainder] + 1;

                if ($dp[$i][$remainder] < $frequency) {
                    $dp[$i][$remainder] = $frequency;
                }

                $maximum = max($maximum, $frequency);
            }
        }

        return $maximum + 1;
    }
}

File: JULY-2025/17-06-1015/p3202.swift
class Solution {
    func maximumLength(_ nums: [Int], _ k: Int) -> Int {
        var dp = Array(repeating: Array(repeating: 0, count: k), count: k)
        var result = 0
        for num in nums {
            let val = num % k
            for i in 0..<k {
                dp[i][val] = dp[val][i] + 1
                result = max(result, dp[i][val])
            }
        }
        return result
    }
}